src/controller_examples/rabbitmq_controller/proof/helper_invariants.rs

```diff
@@ -0,0 +1,627 @@
+    temp_pred_equality(
+        lift_action(stronger_next),
+        lift_action(RMQCluster::next())
+        .and(lift_state(RMQCluster::each_key_in_reconcile_is_consistent_with_its_object()))
+        .and(lift_state(RMQCluster::rest_id_counter_is_no_smaller_than(rest_id)))
+        .and(lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq)))
+    );
+
+    assert forall |s, s_prime: RMQCluster| invariant(s) && #[trigger] stronger_next(s, s_prime)
+    implies invariant(s_prime) by {
+        assert forall |msg: Message|
+            #[trigger] s_prime.network_state.in_flight.contains(msg)
+            && cm_update_request_msg_since(rabbitmq.object_ref(), rest_id)(msg)
+        implies msg.content.get_update_request().obj.spec == ConfigMapView::marshal_spec((make_server_config_map(rabbitmq).data, ())) by {
+            if !s.message_in_flight(msg) {
+                let step = choose |step| RMQCluster::next_step(s, s_prime, step);
+                assert(step.is_ControllerStep());
+                let other_rmq = s.reconcile_state_of(step.get_ControllerStep_0().2.get_Some_0()).triggering_cr;
+                seq_lemmas::seq_equal_preserved_by_add(
+                    other_rmq.metadata.name.get_Some_0(),
+                    rabbitmq.metadata.name.get_Some_0(),
+                    new_strlit("-server-conf")@
+                );
+                assert(other_rmq.object_ref() == rabbitmq.object_ref());
+                assert(other_rmq == s.reconcile_state_of(other_rmq.object_ref()).triggering_cr);
+                assert(rabbitmq.spec() == other_rmq.spec());
+            }
+        }
+    }
+
+    init_invariant(spec, init, stronger_next, invariant);
+}
+
+pub open spec fn cm_delete_request_msg(key: ObjectRef) -> FnSpec(Message) -> bool {
+    |msg: Message|
+        msg.dst.is_KubernetesAPI()
+        && msg.content.is_delete_request()
+        && msg.content.get_delete_request().key == make_server_config_map_key(key)
+}
+
+pub open spec fn cm_delete_request_msg_since(key: ObjectRef, rest_id: RestId) -> FnSpec(Message) -> bool {
+    |msg: Message|
+        cm_delete_request_msg(key)(msg)
+        && msg.content.get_rest_id() >= rest_id
+}
+
+pub open spec fn no_delete_cm_req_since_rest_id_is_in_flight(
+    key: ObjectRef, rest_id: RestId
+) -> StatePred<RMQCluster>
+    recommends
+        key.kind.is_CustomResourceKind(),
+{
+    |s: RMQCluster| {
+        forall |msg: Message| !{
+            &&& #[trigger] s.message_in_flight(msg)
+            &&& cm_delete_request_msg_since(key, rest_id)(msg)
+        }
+    }
+}
+
+// TODO: fix this lemma.
+// If the configmap has no owner_reference, fixing the lemma is simple:
+// we just need to show that the configmap in the cluster state never has any owner reference.
+//
+// However, later we are going to set the owner_reference of the configmap to the CR object,
+// so we will need the assumption that "CR always exists" to prove this invariant.
+#[verifier(external_body)]
```

Comment:
This is left for the next PR?

Reply:
My plan is that if you don't review this PR immediately, I will try to fix those lemmas. Now that you are reviewing it, we can leave it for next PR.

Reply:
OK